### ISSUE TITLE
Solves Model uses Integer type but IP addresses can be a BigInteger

### ISSUE THREAD

**Issue by TheoKlein** [CTFd/CTFd#302]
I follow the steps in [wiki](https://github.com/CTFd/CTFd/wiki/Getting-Started) and I didn't change any default configuration.
Here's the full log when `docker-compose up`:
```
db_1    | Initializing database
ctfd_1  | Waiting on MySQL
db_1    |
db_1    | PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !
db_1    | To do so, start the server, then issue the following commands:
db_1    |
db_1    | '/usr/bin/mysqladmin' -u root password 'new-password'
db_1    | '/usr/bin/mysqladmin' -u root -h  password 'new-password'
db_1    |
db_1    | Alternatively you can run:
db_1    | '/usr/bin/mysql_secure_installation'
db_1    |
db_1    | which will also give you the option of removing the test
db_1    | databases and anonymous user created by default.  This is
db_1    | strongly recommended for production servers.
db_1    |
db_1    | See the MariaDB Knowledgebase at http://mariadb.com/kb or the
db_1    | MySQL manual for more instructions.
db_1    |
db_1    | Please report any problems at http://mariadb.org/jira
db_1    |
db_1    | The latest information about MariaDB is available at http://mariadb.org/.
db_1    | You can find additional information about the MySQL part at:
db_1    | http://dev.mysql.com
db_1    | Consider joining MariaDB's strong and vibrant community:
db_1    | https://mariadb.org/get-involved/
db_1    |
db_1    | Database initialized
db_1    | MySQL init process in progress...
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] mysqld (mysqld 10.2.6-MariaDB-10.2.6+maria~jessie) starting as process 97 ...
db_1    | 2017-07-04  6:48:27 139852247259008 [Warning] Setting lower_case_table_names=2 because file system for /var/lib/mysql/ is case insensitive
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Uses event mutexes
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Compressed tables use zlib 1.2.8
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Using Linux native AIO
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Number of pools: 1
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Using SSE2 crc32 instructions
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Completed initialization of buffer pool
db_1    | 2017-07-04  6:48:27 139851511580416 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
db_1    | 2017-07-04  6:48:27 139852247259008 [Note] InnoDB: Highest supported file format is Barracuda.
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] InnoDB: 128 out of 128 rollback segments are active.
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] InnoDB: Creating shared tablespace for temporary tables
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] InnoDB: Waiting for purge to start
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] InnoDB: 5.7.14 started; log sequence number 1620107
db_1    | 2017-07-04  6:48:28 139851366127360 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
db_1    | 2017-07-04  6:48:28 139851366127360 [Note] InnoDB: Buffer pool(s) load completed at 170704  6:48:28
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] Plugin 'FEEDBACK' is disabled.
db_1    | 2017-07-04  6:48:28 139852247259008 [Warning] 'user' entry 'root@2629d3170cc1' ignored in --skip-name-resolve mode.
db_1    | 2017-07-04  6:48:28 139852247259008 [Warning] 'proxies_priv' entry '@% root@2629d3170cc1' ignored in --skip-name-resolve mode.
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] Reading of all Master_info entries succeded
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] Added new Master_info '' to hash table
db_1    | 2017-07-04  6:48:28 139852247259008 [Note] mysqld: ready for connections.
db_1    | Version: '10.2.6-MariaDB-10.2.6+maria~jessie'  socket: '/var/run/mysqld/mysqld.sock'  port: 0  mariadb.org binary distribution
db_1    | Warning: Unable to load '/usr/share/zoneinfo/leap-seconds.list' as time zone. Skipping it.
db_1    | 2017-07-04  6:48:45 139851595876096 [Warning] 'proxies_priv' entry '@% root@2629d3170cc1' ignored in --skip-name-resolve mode.
db_1    | 2017-07-04  6:48:45 139851595876096 [Warning] 'proxies_priv' entry '@% root@2629d3170cc1' ignored in --skip-name-resolve mode.
db_1    |
db_1    | 2017-07-04  6:48:45 139851450713856 [Note] mysqld (unknown): Normal shutdown
db_1    |
db_1    | 2017-07-04  6:48:45 139851450713856 [Note] Event Scheduler: Purging the queue. 0 events
db_1    | 2017-07-04  6:48:46 139851349341952 [Note] InnoDB: FTS optimize thread exiting.
db_1    | 2017-07-04  6:48:46 139851450713856 [Note] InnoDB: Starting shutdown...
db_1    | 2017-07-04  6:48:46 139851366127360 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
db_1    | 2017-07-04  6:48:46 139851366127360 [Note] InnoDB: Buffer pool(s) dump completed at 170704  6:48:46
db_1    | 2017-07-04  6:48:47 139851450713856 [Note] InnoDB: Shutdown completed; log sequence number 1620135
db_1    | 2017-07-04  6:48:47 139851450713856 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
db_1    | 2017-07-04  6:48:47 139851450713856 [Note] mysqld: Shutdown complete
db_1    |
db_1    |
db_1    | MySQL init process done. Ready for start up.
db_1    |
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] mysqld (mysqld 10.2.6-MariaDB-10.2.6+maria~jessie) starting as process 1 ...
db_1    | 2017-07-04  6:48:47 140170890749824 [Warning] Setting lower_case_table_names=2 because file system for /var/lib/mysql/ is case insensitive
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] InnoDB: Uses event mutexes
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] InnoDB: Compressed tables use zlib 1.2.8
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] InnoDB: Using Linux native AIO
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] InnoDB: Number of pools: 1
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] InnoDB: Using SSE2 crc32 instructions
db_1    | 2017-07-04  6:48:47 140170890749824 [Note] InnoDB: Initializing buffer pool, total size = 256M, instances = 1, chunk size = 128M
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: Completed initialization of buffer pool
db_1    | 2017-07-04  6:48:48 140170151692032 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: Highest supported file format is Barracuda.
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: 128 out of 128 rollback segments are active.
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: Creating shared tablespace for temporary tables
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: Waiting for purge to start
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] InnoDB: 5.7.14 started; log sequence number 1620135
db_1    | 2017-07-04  6:48:48 140170007406336 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
db_1    | 2017-07-04  6:48:48 140170007406336 [Note] InnoDB: Buffer pool(s) load completed at 170704  6:48:48
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] Plugin 'FEEDBACK' is disabled.
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] Server socket created on IP: '::'.
db_1    | 2017-07-04  6:48:48 140170890749824 [Warning] 'proxies_priv' entry '@% root@2629d3170cc1' ignored in --skip-name-resolve mode.
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] Reading of all Master_info entries succeded
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] Added new Master_info '' to hash table
db_1    | 2017-07-04  6:48:48 140170890749824 [Note] mysqld: ready for connections.
db_1    | Version: '10.2.6-MariaDB-10.2.6+maria~jessie'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
db_1    | 2017-07-04  6:48:49 140170583656192 [Warning] Access denied for user 'root'@'172.21.0.3' (using password: NO)
ctfd_1  | ............................Ready
ctfd_1  | Starting CTFd
db_1    | 2017-07-04  6:48:51 140170583656192 [Warning] Aborted connection 9 to db: 'unconnected' user: 'ctfd' host: '172.21.0.3' (Got an error reading communication packets)
db_1    | 2017-07-04  6:54:20 140170553988864 [Warning] Aborted connection 11 to db: 'ctfd' user: 'ctfd' host: '172.21.0.3' (Got an error reading communication packets)
```
It's strange that admin can add challenge but no one can submit the key successfully include admin.

**Comment by romanSil** [comment]
What does `SELECT Host, User FROM mysql.user;` say?

**Comment by TheoKlein** [comment]
Hi @romanSil , here's the result:
```
MariaDB [(none)]> SELECT Host, User FROM mysql.user;
+-----------+------+
| Host      | User |
+-----------+------+
| %         | ctfd |
| %         | root |
| localhost | root |
+-----------+------+
```

**Comment by wirabdillah** [comment]
I met the same problem. Here's my error.log

> [2017-07-08 09:18:38,834] ERROR in app: Exception on /chal/1 [POST]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/opt/CTFd/CTFd/challenges.py", line 299, in chal
    db.session.commit()
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/scoping.py", line 157, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 874, in commit
    self.transaction.commit()
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 461, in commit
    self._prepare_impl()
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 441, in _prepare_impl
    self.session.flush()
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2139, in flush
    self._flush(objects)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2259, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2223, in _flush
    flush_context.execute()
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 389, in execute
    rec.execute(self)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 548, in execute
    uow
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 181, in save_obj
    mapper, table, insert)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 835, in _emit_insert_statements
    execute(statement, params)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 945, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
    context)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1393, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
    context)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
    cursor.execute(statement, parameters)
  File "/usr/local/lib/python2.7/site-packages/pymysql/cursors.py", line 166, in execute
    result = self._query(query)
  File "/usr/local/lib/python2.7/site-packages/pymysql/cursors.py", line 322, in _query
    conn.query(q)
  File "/usr/local/lib/python2.7/site-packages/pymysql/connections.py", line 852, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/usr/local/lib/python2.7/site-packages/pymysql/connections.py", line 1053, in _read_query_result
    result.read()
  File "/usr/local/lib/python2.7/site-packages/pymysql/connections.py", line 1336, in read
    first_packet = self.connection._read_packet()
  File "/usr/local/lib/python2.7/site-packages/pymysql/connections.py", line 1010, in _read_packet
    packet.check_error()
  File "/usr/local/lib/python2.7/site-packages/pymysql/connections.py", line 393, in check_error
    err.raise_mysql_exception(self._data)
  File "/usr/local/lib/python2.7/site-packages/pymysql/err.py", line 107, in raise_mysql_exception
    raise errorclass(errno, errval)
DataError: (pymysql.err.DataError) (1264, u"Out of range value for column 'ip' at row 1") [SQL: u'INSERT INTO solves (chalid, teamid, ip, flag, date) VALUES (%(chalid)s, %(teamid)s, %(ip)s, %(flag)s, %(date)s)'] [parameters: {'date': datetime.datetime(2017, 7, 8, 9, 18, 38, 832019), 'ip': 2886860801, 'flag': u'CTF{all1zw3ll}', 'teamid': 2, 'chalid': 1}]

Column ip is inserted with 2886860801 which int type cannot handle

**Comment by ColdHeat** [comment]
@wirabdillah I think that is a different issue from the one here, can you make a new issue?

**Comment by ColdHeat** [comment]
Actually, I take that back @TheoKlein can you be more specific about your problem? Are you having trouble starting CTFd or are you having trouble submitting a flag?

**Comment by TheoKlein** [comment]
My problem is submitting a flag but I only focus on docker log. Although there are some warnings, CTFd can startup without any trouble.

@wirabdillah 's report is right, I forgot to check `error.log`. This issue should be closed. Thanks for your answer!

**Comment by ColdHeat** [comment]
Ok I think I get what you're saying. Regardless I'm fixing the issue now.

**Comment by ColdHeat** [comment]
I'm closing this in favor of a new issue as the problem is actually somewhat unnecessarily deep. Ultimately I think CTFd will need to store IP addresses as strings or leverage the SQLAlchemy-Utils IPaddress type.

**Comment by ColdHeat** [comment]
I'll be fixing this issue in a PR but https://github.com/CTFd/CTFd/issues/306 is the real issue.

**Comment by romanSil** [comment]
I also had the issue with submitting flags. Solved by changing the SQL column to unsigned.

Is it useful ti save the IP when using docker? I have only docker intern IP addresses in there.
